fulp_modules/devils/devil_powers.dm

```diff
@@ -0,0 +1,160 @@
+	invocation = "I ain't had this much fun since Georgia."
+	action_icon_state = "golden_violin"
+	name = "Summon golden violin"
+	action_icon = 'fulp_modules/devils/devil_icon/devil_icons.dmi'
+	action_background_icon_state = "bg_demon"
+
+/obj/item/instrument/violin/golden/devil
+	name = "golden violin"
+	desc = "A golden musical instrument with four strings and a bow. \"The devil went down to space, he was looking for an assistant to grief.\""
+	icon = 'fulp_modules/devils/devil_icon/violin.dmi'
+	lefthand_file = 'fulp_modules/devils/devil_icon/devils_lefthand.dmi'
+	righthand_file = 'fulp_modules/devils/devil_icon/devils_righthand.dmi'
+	icon_state = "golden_violin"
+	inhand_icon_state = "golden_violin"
+	resistance_flags = LAVA_PROOF | FIRE_PROOF | ACID_PROOF
+
+/// Summon Fireball
+/obj/effect/proc_holder/spell/aimed/fireball/hellish
+	name = "Hellfire"
+	desc = "This spell launches hellfire at the target."
+
+	school = "evocation"
+	charge_max = 80
+	clothes_req = FALSE
+	invocation = "Your very soul will catch fire!"
+	invocation_type = INVOCATION_SHOUT
+	range = 2
+
+	projectile_type = /obj/projectile/magic/aoe/fireball/infernal
+
+	action_background_icon_state = "bg_demon"
+
+/obj/projectile/magic/aoe/fireball/infernal
+	name = "infernal fireball"
+	exp_heavy = -1
+	exp_light = -1
+	exp_flash = 4
+	exp_fire= 5
+
+/obj/projectile/magic/aoe/fireball/infernal/on_hit(target)
+	. = ..()
+	if(ismob(target))
+		var/mob/living/M = target
+		if(M.anti_magic_check())
+			return BULLET_ACT_BLOCK
+	var/turf/T = get_turf(target)
+	for(var/i=0, i<50, i+=10)
+		addtimer(CALLBACK(GLOBAL_PROC, .proc/explosion, T, -1, exp_heavy, exp_light, exp_flash, FALSE, FALSE, exp_fire), i)
+
+
+/// Summon Dancefloor
+/obj/effect/proc_holder/spell/targeted/summon_dancefloor
+	name = "Summon Dancefloor"
+	desc = "When what a Devil really needs is funk."
+	include_user = TRUE
+	range = -1
+	clothes_req = FALSE
+
+	school = "conjuration"
+	charge_max = 10
+	cooldown_min = 50 //5 seconds, so the smoke can't be spammed
+	action_icon = 'fulp_modules/devils/devil_icon/devil_icons.dmi'
+	action_icon_state = "funk"
+	action_background_icon_state = "bg_demon"
+
+	var/list/dancefloor_turfs
+	var/list/dancefloor_turfs_types
+	var/dancefloor_exists = FALSE
+	var/datum/effect_system/smoke_spread/transparent/dancefloor_devil/smoke
+
+/obj/effect/proc_holder/spell/targeted/summon_dancefloor/cast(list/targets, mob/user = usr)
+	LAZYINITLIST(dancefloor_turfs)
+	LAZYINITLIST(dancefloor_turfs_types)
+
+	if(!smoke)
+		smoke = new()
+	smoke.set_up(0, get_turf(user))
+	smoke.start()
+
+	if(dancefloor_exists)
+		dancefloor_exists = FALSE
+		for(var/i in 1 to dancefloor_turfs.len)
+			var/turf/T = dancefloor_turfs[i]
+			T.ChangeTurf(dancefloor_turfs_types[i], flags = CHANGETURF_INHERIT_AIR)
+	else
+		var/list/funky_turfs = RANGE_TURFS(1, user)
+		for(var/turf/closed/solid in funky_turfs)
+			to_chat(user, "<span class='warning'>You're too close to a wall.</span>")
+			return
```

Comment:
can't you just... skip dance floor placements on tiles that have walls, instead of making the entire spell fail if you're near a wall?